### PR TITLE
Add NSS endpoints for ECH interop

### PR DESCRIPTION
**Pull request by kjacobs-moz**
This currently works for `ech-accept` between NSS and cloudflare-go.

### CODE REVIEW

**Comment by cjpatton** [comment]
Is this ready to review, @kjacobs-moz?

**Comment by kjacobs-moz** [comment]
> Is this ready to review, @kjacobs-moz?

Yes, the NSS change has landed.

File: impl-endpoints/nss/ech_key_converter.py
Comment (cjpatton) [inline impl-endpoints/nss/ech_key_converter.py:13]: Suggest s/ech_keys/ech_keypair/. ech_keys sounds like it's a sequence of ECHKey structures.

File: impl-endpoints/nss/run_endpoint.sh
Comment (cjpatton) [inline impl-endpoints/nss/run_endpoint.sh:20]: nit: Align these lines with the indentation of the block.

File: impl-endpoints/nss/ech_key_converter.py
Comment (cjpatton) [inline impl-endpoints/nss/ech_key_converter.py:8]: How does NSS Interpret these values? Are they specific to X25519? If so, please add a comment explaining that they are.

File: cmd/util/make.go
Comment (cjpatton) [inline cmd/util/make.go:109]: <3

**Review by cjpatton — CHANGES_REQUESTED** [review]
This looks great! Just a few minor nits and one ask (see inline comments).

Also, a question: The ECH draft does not specify how to format secret keys. Would you endorse PKCS#8 as a format for our purposes here? In other words, should we modify `util.go` to output an ECHConfigs and, separately, an the corresponding PKCS#8-formatted key pair? (See #17.)

**Comment by cjpatton** [comment]
> For the other direction, I'll need to land a change to [selfserv.c](https://searchfox.org/mozilla-central/source/security/nss/cmd/selfserv/selfserv.c#1967) to consume the PKCS8-format HPKE keypair. With that change applied, connection succeeds from cloudflare-go to NSS.

FWIW, cloudflare-go client -> nss server works correctly with the existing conversion script.

**Review by chris-wood — APPROVED** [review]
LGTM pending @cjpatton's suggestions for the key conversion script.

**Comment by cjpatton** [comment]
> LGTM pending @cjpatton's suggestions for the key conversion script.

I'm happy to merge without the conversion and get back to it later.

**Comment by chris-wood** [comment]
Yeah, same. I was referring to the documentation suggestions. But even those I'd be fine without.

**Comment by kjacobs-moz** [comment]
> > LGTM pending @cjpatton's suggestions for the key conversion script.
> 
> I'm happy to merge without the conversion and get back to it later.

Thanks for the reviews.

Unfortunately, I'm not going to have time in the immediate future to rewrite it in Go, but I would definitely support PKCS8-formatting the ECH/HPKE keypair rather than storing the raw private key, which some libraries may not be able to import easily. NSS is one such library, and doing it outside of NSS/selfserv avoids having to write a second ECHConfigs parser at the application level. With that change, we could remove the script entirely, but it might be worth waiting to see if other libraries have an opinion.

Another option is to output two formats from util.go.

**Review by cjpatton — APPROVED** [review]
LGTM. Pleas squash the last commit before merging.

**Comment by cjpatton** [comment]
> LGTM. Pleas squash the last commit before merging.

Oops, scratch that. Commits look good. I'm merging now.

**Pull request merged**
